src/Language/PureScript/TypeChecker/Deriving.hs

```diff
@@ -359,6 +368,128 @@ decomposeRec' = sortOn fst . go
   where go (RCons _ str typ typs) = (str, typ) : go typs
         go _ = []
 
+data ParamUsage
+  = IsParam
+  | MentionsParam ParamUsage
+  | IsRecord (NonEmpty (PSString, ParamUsage))
+
+type ParamErrorData = [([Either (ProperName 'ConstructorName) PSString], SourceSpan)]
+
+validateParamsInTypeConstructors
+  :: forall m
+   . MonadError MultipleErrors m
+  => MonadState CheckState m
+  => ModuleName
+  -> ProperName 'TypeName
+  -> m [(ProperName 'ConstructorName, [Maybe ParamUsage])]
+validateParamsInTypeConstructors mn tyConNm = do
+  (_, _, tyArgNames, ctors) <- lookupTypeDecl mn tyConNm
+  param <- note (errorMessage $ KindsDoNotUnify (kindType -:> kindType) kindType) . lastMay $ map fst tyArgNames
+  ctors' <- traverse (traverse $ traverse replaceAllTypeSynonyms) ctors
```

Comment:
I think these lines are what causes the following code to not work. 
```purs
type Rec2 a f =
  { foo :: f a }
  
data Test f a = Test1 (Rec2 a f)
derive instance Foldable f => Foldable (Test f)
```
Right now, you don't replace the type synonyms before checking the kind of the last argument. The advantage of this is that you don't get the duplicate record labels in cases like `FoldableInstance9.purs`' `Test6` (see the second part of https://github.com/purescript/purescript/pull/4371#issuecomment-1242793472). However, it also means the above code produces the following error message:
```
Error found:
       in module FoldableInstance9
       at tests/purs/failing/FoldableInstance9.purs:97:19 - 97:20 (line 97, column 19 - line 97, column 20)
       
         Could not match kind
                               
           Type -> Type -> Type
                               
         with kind
               
           Type
               
       
       while checking that type f
         has kind Type -> Type
       while inferring the kind of Rec2 a f
       in type constructor Test
       
       See https://github.com/purescript/documentation/blob/master/errors/KindsDoNotUnify.md for more information,
       or to contribute content related to this error.
       

  To rerun use: --match "/compiler/Failing examples/'FoldableInstance9.purs' should fail to compile/"
```

I would argue that this example should also be derivable. Thoughts?

Reply:
> I'm not getting any errors when I try it.

Ah! You're right. Sorry. The kinds-do-not-unify error I was getting was because the example I tested didn't include a `Foldable f` constraint. I added a `Test8` for this in the `FoldableInstance9.purs` file, found it didn't work, and simplified it to the above. But I didn't test the above separately.